src/main/java/com/yelp/nrtsearch/ArchiverModule.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright 2021 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.AnonymousAWSCredentials;
+import com.amazonaws.auth.profile.ProfileCredentialsProvider;
+import com.amazonaws.auth.profile.ProfilesConfigFile;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.transfer.TransferManagerBuilder;
+import com.google.inject.*;
+import com.google.inject.name.Named;
+import com.google.inject.name.Names;
+import com.yelp.nrtsearch.server.backup.*;
+import com.yelp.nrtsearch.server.config.LuceneServerConfiguration;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ArchiverModule extends AbstractModule {
+  private static final Logger logger = LoggerFactory.getLogger(ArchiverModule.class);
+
+  @Override
+  protected void configure() {
+    bind(ContentDownloader.class)
+        .annotatedWith(Names.named("contentDownloaderNoTar"))
+        .toProvider(ContentDownloaderNoTar.class)
+        .asEagerSingleton();
+    bind(ContentDownloader.class)
+        .annotatedWith(Names.named("contentDownloaderWithTar"))
+        .toProvider(ContentDownloaderWithTar.class)
+        .asEagerSingleton();
+    bind(FileCompressAndUploader.class)
+        .annotatedWith(Names.named("fileCompressAndUploaderNoTar"))
+        .toProvider(FileCompressAndUploaderNoTar.class)
+        .asEagerSingleton();
+    bind(FileCompressAndUploader.class)
+        .annotatedWith(Names.named("fileCompressAndUploaderWithTar"))
+        .toProvider(FileCompressAndUploaderWithTar.class)
+        .asEagerSingleton();
+    bind(Archiver.class)
+        .annotatedWith(Names.named("legacyArchiver"))
+        .toProvider(LegacyArchiverProvider.class)
+        .asEagerSingleton();
+    bind(BackupDiffManager.class).toProvider(BackupDiffManagerProvider.class).asEagerSingleton();
+    bind(Archiver.class)
+        .annotatedWith(Names.named("incArchiver"))
+        .toProvider(IncArchiverProvider.class)
+        .asEagerSingleton();
+  }
```

Comment:
[NamedAnnotation](https://github.com/google/guice/wiki/BindingAnnotations) allows us to bind multiple instances for the same type e.g. "fileCompressAndUploaderNoTar"  and "fileCompressAndUploaderWithTar"  both produce instances of class `FileCompressAndUploader` but one uses `NoTarImpl` while other uses `TarImpl`. 

Same thing with impl of `Archiver` being named `legacyArchiver` or not.

Reply:
You can still use a method annotated by `Provides` and annotate it with `Named` to bind multiple instances for the same type like this:
```
@Named("legacyArchiver")
@Inject
@Singleton
@Provides
public Tar providesLegacyArchiver(AmazonS3 s3, LuceneServerConfiguration luceneServerConfiguration, Tar tar) {
Path archiveDir = Paths.get(luceneServerConfiguration.getArchiveDirectory());
      return new ArchiverImpl(
          s3,
          luceneServerConfiguration.getBucketName(),
          archiveDir,
          tar,
          luceneServerConfiguration.getDownloadAsStream());
}
```

Reply:
So you are suggesting usage of @Provides annotation while current code implements a Provider like so:

```
private static class LegacyArchiverProvider implements Provider<Archiver> {
    @Inject AmazonS3 s3;
    @Inject LuceneServerConfiguration luceneServerConfiguration;
    @Inject Tar tar;

    @Override
    public Archiver get() {
      Path archiveDir = Paths.get(luceneServerConfiguration.getArchiveDirectory());
      return new ArchiverImpl(
          s3,
          luceneServerConfiguration.getBucketName(),
          archiveDir,
          tar,
          luceneServerConfiguration.getDownloadAsStream());
    }
```

and this is how we bind it:
```
    bind(Archiver.class)
        .annotatedWith(Names.named("legacyArchiver"))
        .toProvider(LegacyArchiverProvider.class)
        .asEagerSingleton();

```

I am not sure that one approach is definitively superior over other? IMO its a styling issue, details on differences here: https://stackoverflow.com/questions/28202882/guice-provides-methods-vs-provider-classes.

Reply:
From the stackoverflow link: `The main advantage of @Provides is compactness, especially in comparison to anonymous inner Provider implementations.`
My main concern though is using only one style instead of mixing two. The Provides annotation approach is already used elsewhere too so I'd favor that.

Reply:
Yes I read the link before sending over :) . The Provider implementation feels more powerful to me as the link suggests as well, but I have no opinion in the usages here, since its mostly a styling issue here.

Reply:
As I said my main concern is that we are mixing two styles and I would prefer to use one.

Reply:
As discussed tried using Named annotation instead of Provider implementation. Could not get it to work due to Missing implementation guice error: https://gist.github.com/umeshdangat/4343274949a221ed5832bed9a1213b7d. As we spoke in person keeping my current Provider Implementation as is.